src/relations/mysql.py

```diff
@@ -57,6 +57,56 @@ def _get_or_set_password_in_peer_databag(self, username: str) -> str:
 
         return password
 
+    def _get_or_generate_username(self, event_relation_id: int) -> str:
+        """Retrieve username from config or generate a new one.
+
+        Then store it in the peer databag. Assumes that the caller is a leader unit.
+        """
+        username_in_databag = self.charm.app_peer_data.get("mysql-interface-user")
+        username_in_config = self.charm.config.get("mysql-interface-user")
+
+        if username_in_databag:
+            if (
+                username_in_config
+                and username_in_databag != username_in_config
+                and not self.charm._mysql.does_mysql_user_exist(username_in_databag, "%")
+            ):
+                self.charm.app_peer_data["mysql-interface-user"] = username_in_config
+                return username_in_config
+
+            return username_in_databag
+
+        if username_in_config:
+            self.charm.app_peer_data["mysql-interface-user"] = username_in_config
+            return username_in_config
+
+        generated_username = f"relation-{event_relation_id}"
+        self.charm.app_peer_data["mysql-interface-user"] = generated_username
+        return generated_username
+
+    def _get_or_generate_database(self, event_relation_id: int) -> str:
+        """Retrieve database from config or generate a new one.
+
+        Then store it in the peer databag. Assumes that the caller is a leader unit.
+        """
+        database_in_databag = self.charm.app_peer_data.get("mysql-interface-database")
+        database_in_config = self.charm.config.get("mysql-interface-database")
+
+        if database_in_databag:
+            if database_in_databag != database_in_config:
+                self.charm.app_peer_data["mysql-interface-database"] = database_in_config
+                return database_in_config
+
+            return database_in_databag
+
+        if database_in_config:
+            self.charm.app_peer_data["mysql-interface-database"] = database_in_config
+            return database_in_config
+
+        generated_database = f"database-{event_relation_id}"
+        self.charm.app_peer_data["mysql-interface-database"] = generated_database
+        return generated_database
```

Comment:
Sorry for the confusion. When I say "lock", I mean in respect to other relations being formed. I envisioned keeping same the username/database combination the same after being autogenerated or derived from the config, until all relations are broken or the config is changed (in which case the combination should be updated for all relations). Does this make sense/feel right?

Reply:
Pre-defined defaults in config.yaml may lead to two apps will write the same db if related via legacy interface.
It can suprise users more, comparing to new DB-name on re-relation.

IMHO, resolved.

Reply:
I'll try to summarize the behavior as I understand it: 

- if there are no config options set, then the username/database will be autogenerated. if config options set, the config option values will be used
- all relations with the interface will receive the same username/database
- if config changes while relations exist, the new user name/database are created and every relation databag is updated

Reply:
it's a bit more complicated. we need to autogenerate database names (notice we are not deleting databases upon relation-broken). we want to preserve the data when this happens

if we use a default value, we would be presenting the same database upon re-relations (which may not be the desired behavior) -> this was the primary reason why we autogenerate

Reply:
no, it was a requirement by one of the consumers for this legacy relation that when there are more than one relations, all relations get the same username/database (as they have multiple applications writing to the same database)